### PR TITLE
[QA-1926] Run Terra-UI tests on PR commits against Staging backend

### PR DESCRIPTION
**Pull request by aweng98**
[QA-1926](https://broadworkbench.atlassian.net/browse/QA-1926)

New CircleCI workflow diagram

Screen Shot 2022-06-30 at 11 44 46 AM

### CODE REVIEW

**Comment by aweng98** [comment]
Change is not needed [here](https://github.com/DataBiosphere/terra-ui/blob/d4533d8e90a2bc1f947da99c9bf2339222b32191/.circleci/config.yml#L149), right?
`- run: cp config/dev.json build/config.json`

**Review by breilly2 — APPROVED** [review]
👍

**Review by nawatts — CHANGES_REQUESTED** [review]
> Change is not needed [here](https://github.com/DataBiosphere/terra-ui/blob/d4533d8e90a2bc1f947da99c9bf2339222b32191/.circleci/config.yml#L149), right?
> \- run: cp config/dev.json build/config.json

Yes, that line needs to change. Currently, when the `local` parameter is set to true, as it is in the `integration-tests-branch-against-staging` job, the local UI will be pointed at the dev backend.

File: .circleci/config.yml
Comment (nawatts) [inline .circleci/config.yml]: The env here should stay as `local`. Changing it to `staging` runs tests against the UI deployed on staging, not the local UI in the branch.

Comment (aweng98) [reply .circleci/config.yml]: You're right. Purpose of `local` and `env` were a bit confusing for me.

**Review by nawatts — APPROVED** [review]

**Comment by aweng98** [comment]
FYI, tests are mostly broken with changes. I will troubleshoot.

File: integration-tests/utils/terra-envs.js
Comment (nawatts) [inline integration-tests/utils/terra-envs.js]: `testUrl` should stay as `http://localhost:3000` for the local environment. This might explain several test failures if tests in this branch (which is several commits behind the dev branch) are being run against the UI deployed on staging.

**Comment by nawatts** [comment]
Noticed at least one test where the test code assumes that a local UI is pointed at dev.
https://github.com/DataBiosphere/terra-ui/blob/adb4d9a1bdbc3a9d79033a2a63830b613665e302/integration-tests/tests/find-workflow.js#L32-L37

**Review by nawatts — CHANGES_REQUESTED** [review]
Requesting changes for the `testUrl` in terra-envs.js and the find-workflow code noted above.

**Comment by aweng98** [comment]
Made changes but still seening some fail. Looking...
Also found couple places where `dev` is hardcoded: [workspace-dashboard.js#L161](https://github.com/DataBiosphere/terra-ui/blob/adb4d9a1bdbc3a9d79033a2a63830b613665e302/integration-tests/tests/workspace-dashboard.js#L161) and [preview-drs-uri.js#L17](https://github.com/DataBiosphere/terra-ui/blob/adb4d9a1bdbc3a9d79033a2a63830b613665e302/integration-tests/tests/preview-drs-uri.js#L17)
What do I do?

**Comment by kyuksel** [comment]
> Made changes but still seening some fail. Looking... Also found couple places where `dev` is hardcoded: [workspace-dashboard.js#L161](https://github.com/DataBiosphere/terra-ui/blob/adb4d9a1bdbc3a9d79033a2a63830b613665e302/integration-tests/tests/workspace-dashboard.js#L161) and [preview-drs-uri.js#L17](https://github.com/DataBiosphere/terra-ui/blob/adb4d9a1bdbc3a9d79033a2a63830b613665e302/integration-tests/tests/preview-drs-uri.js#L17) What do I do?

Those seem to mock API responses. @cahrens do they need to specify an environment?

**Comment by aweng98** [comment]
`run-analysis` and `run-notebook` tests are not working for local UI pointed to Staging backend. Error in popup is `Error setting up notebook`. It seems happen every time.

**Comment by nawatts** [comment]
> run-analysis and run-notebook tests are not working for local UI pointed to Staging backend. Error in popup is Error setting up notebook. It seems happen every time.

The notebook UI is served by Leonardo and embedded in an iframe in terra-ui. I suspect dev Leonardo's content security policy allows http://localhost:3000 to load it in an iframe and staging Leonardo's does not.

File: integration-tests/tests/run-catalog-filter.js
Comment (aweng98) [inline integration-tests/tests/run-catalog-filter.js:61]: This test is one of four tests that don't run against Staging backend. Other three tests seem work just fine. Jira ticket to fix all four tests to run against Staging backend is [QA-1930](https://broadworkbench.atlassian.net/browse/QA-1930)

**Comment by aweng98** [comment]
Reached out to IA team for help with notebook test failures. This PR is currently blocked.

**Comment by aweng98** [comment]
Since two problematic notebook tests (`run-notebook.js` and `run-analysis.js`) are disabled for `local` and `dev`, PR commit test is passing. No test failures. PTAL.

**Comment by nawatts** [comment]
Given that we previously held off on this change because of failing notebook tests (https://github.com/DataBiosphere/terra-ui/pull/3191#issuecomment-1176349235), I think we should wait until those tests are re-enabled and we're sure they work with the local UI pointed at staging.

**Comment by aweng98** [comment]
Can I convince you to reconsider? I suggest we don't wait. 
This PR allows us to see any improvement to the tests stablity on PR commits. Team Leo is the owner of all notebook tests and is responsible for fixing problems in tests if problems are not caused by UI test framework. 
Leo team is well suited to the task of fixing notebook tests. PR that re-enables tests cannot merge if notebook tests don't run and pass on commits.

**Comment by nawatts** [comment]
We've [disabled notebook tests on PRs](https://github.com/DataBiosphere/terra-ui/pull/3213) due to what appears to be an issue with the Leonardo service.

Prior to that, there seemed to be some other issue that prevented notebook tests from running successfully with a local UI pointed at staging (https://github.com/DataBiosphere/terra-ui/pull/3191#issuecomment-1176349235). When notebook tests were enabled, that issue blocked this PR.

Thus, I think making this change now would add another obstacle to getting notebook tests running on PRs again.

**Comment by aweng98** [comment]
Asked DC team about `run-catalog-filter.js` test failure [here](https://broadworkbench.atlassian.net/browse/DC-354). This test failure is not caused by this PR.

**Review by nawatts — COMMENTED** [review]
> Asked DC team about run-catalog-filter.js test failure [here](https://broadworkbench.atlassian.net/browse/DC-354). This test failure is not caused by this PR.

This looks like it's related to different data on dev and staging.

The test failed with this error:
> TimeoutError: waiting for XPath `//*[contains(normalize-space(.),"acoustic neuroma")]` failed: timeout 30000ms exceeded

Compare https://bvdp-saturn-dev.appspot.com/#library/datasets to https://bvdp-saturn-staging.appspot.com/#library/datasets. "acoustic neuroma" shows up in the "Disease" section of the sidebar on dev, but not on staging.

run-catalog-filter.js is currently configured to run on local and dev environments. 
https://github.com/DataBiosphere/terra-ui/blob/edd7b41492491e9e9ab83476feef1f3994f9e4d1/integration-tests/tests/run-catalog-filter.js#L126

This changes the local environment from dev to staging. With this change, I would expect run-catalog-filter to consistently fail when run on PRs.

**Comment by aweng98** [comment]
Yeah, there are 4 data-catalog UI tests which I have identified early on don't run on Staging. [QA-1930](https://broadworkbench.atlassian.net/browse/QA-1930). What I meant with `This test failure is not caused by this PR` is these test failures are test issues.

**Comment by tcjiang9** [comment]
re: the data being different between dev and staging, that shouldn't matter because I changed the test to return mock data and operate on that. Perhaps the system being used to mock return data only works in dev?
I'm doing

```
const setDatasetsMockValues = async page => {
  const datasetsResult = {
}
  return await page.evaluate(datasetsResult => {
    window.ajaxOverridesStore.set([
      {
        filter: { url: new RegExp(`/api/v1/datasets(.*)`, 'g') },
        fn: () => () => {
          return Promise.resolve(new Response(JSON.stringify(datasetsResult), { status: 200 }))
        }
      }
    ])
  }, datasetsResult)
```

**Comment by tcjiang9** [comment]
@nawatts re: the data being different  between dev and staging, I changed the test to use mock data and operate on that mock data, perhaps the system I'm using to mock our API result doesn't work outside of dev? I'm doing

```
const setDatasetsMockValues = async page => {
  const datasetsResult = {
// mock values here
}
  return await page.evaluate(datasetsResult => {
    window.ajaxOverridesStore.set([
      {
        filter: { url: new RegExp(`/api/v1/datasets(.*)`, 'g') },
        fn: () => () => {
          return Promise.resolve(new Response(JSON.stringify(datasetsResult), { status: 200 }))
        }
      }
    ])
  }, datasetsResult)
}

```

**Comment by nawatts** [comment]
It looks like the test navigates to the Datasets page (and thus loads datasets) before installing the mock responses. So I suspect the mock responses are not having an effect in any environment but the test works with the data that is on dev.

https://github.com/DataBiosphere/terra-ui/blob/edd7b41492491e9e9ab83476feef1f3994f9e4d1/integration-tests/tests/run-catalog-filter.js#L18-L19

**Comment by aweng98** [comment]
> > Made changes but still seening some fail. Looking... Also found couple places where `dev` is hardcoded: [workspace-dashboard.js#L161](https://github.com/DataBiosphere/terra-ui/blob/adb4d9a1bdbc3a9d79033a2a63830b613665e302/integration-tests/tests/workspace-dashboard.js#L161) and [preview-drs-uri.js#L17](https://github.com/DataBiosphere/terra-ui/blob/adb4d9a1bdbc3a9d79033a2a63830b613665e302/integration-tests/tests/preview-drs-uri.js#L17) What do I do?
> 
> Those seem to mock API responses. @cahrens do they need to specify an environment?

@cahrens Hi, can you help with this test question?

**Comment by nawatts** [comment]
> Yeah, there are 4 data-catalog UI tests which I have identified early on don't run on Staging. [QA-1930](https://broadworkbench.atlassian.net/browse/QA-1930). What I meant with This test failure is not caused by this PR is these test failures are test issues.

It looks like the root cause of the run-catalog-filter test failure is due to that test's reliance on environment-specific data. However, the run-catalog-filter test currently passes on PRs (because it is run against the dev environment) and with this change, it will fail (because it is run against the staging environment). So I would say that that test failure _is_ related to this PR and needs to be resolved before this PR can be merged.

**Comment by aweng98** [comment]
again, move PR to blocked. Awaiting `run-catalog-filter.js` test fix from DC team.

**Comment by aweng98** [comment]
@petesantos FYI. this PR is blocked again.

**Comment by cahrens** [comment]
> > Yeah, there are 4 data-catalog UI tests which I have identified early on don't run on Staging. [QA-1930](https://broadworkbench.atlassian.net/browse/QA-1930). What I meant with This test failure is not caused by this PR is these test failures are test issues.
> 
> It looks like the root cause of the run-catalog-filter test failure is due to that test's reliance on environment-specific data. However, the run-catalog-filter test currently passes on PRs (because it is run against the dev environment) and with this change, it will fail (because it is run against the staging environment). So I would say that that test failure _is_ related to this PR and needs to be resolved before this PR can be merged.

I'm confused… if the tests are disabled on staging, why are they being run against staging as part of PR runs? It seems like they shouldn't be run.

**Comment by nawatts** [comment]
> if the tests are disabled on staging, why are they being run against staging as part of PR runs? It seems like they shouldn't be run.

PR tests run with the environment set to "local". Unfortunately, the environments code treats "local" as another environment like dev, alpha, staging. But that's not really true, since only the UI runs locally and API requests are sent to one of the other environments. Currently, in the dev branch, local UIs in PR tests use dev APIs. But this PR changes them to use staging APIs.

**Comment by aweng98** [comment]
All tests should run agnostic of environment and we don't want to lose any test coverage after switch to using Staging backend for "local".

**Comment by cahrens** [comment]
Thanks, I forgot about "local".

**Comment by nawatts** [comment]
Has #3340 unblocked this?

**Comment by aweng98** [comment]
> Has #3340 unblocked this?

Thank you for the notice. I rebased and ran test locally against Staging. It worked. So changing PR ready for review.

**Comment by aweng98** [comment]
> > Made changes but still seening some fail. Looking... Also found couple places where `dev` is hardcoded: [workspace-dashboard.js#L161](https://github.com/DataBiosphere/terra-ui/blob/adb4d9a1bdbc3a9d79033a2a63830b613665e302/integration-tests/tests/workspace-dashboard.js#L161) and [preview-drs-uri.js#L17](https://github.com/DataBiosphere/terra-ui/blob/adb4d9a1bdbc3a9d79033a2a63830b613665e302/integration-tests/tests/preview-drs-uri.js#L17) What do I do?
> 
> Those seem to mock API responses. @cahrens do they need to specify an environment?

@cahrens Hi, sorry for the re-asking because I don't find your answer to this question. Can you check it out?

**Comment by nawatts** [comment]
Given that both of those tests have been running successfully against alpha and staging in tests on the dev branch, I don't think they require any changes. The tests that did require changes were because they had specific logic for when the test URL is localhost / the environment is "local".

File: .circleci/config.yml
Comment (nawatts) [inline .circleci/config.yml:334]: Should we keep the tests against staging in nightly? Is this meant to run tests after updating the staging backend?

Comment (aweng98) [reply .circleci/config.yml:334]: There was a previous discussion to remove it from nightly because we run tests against Staging API on PR commits and against Staging UI + API when merges.

**Comment by cahrens** [comment]
> > > Made changes but still seening some fail. Looking... Also found couple places where `dev` is hardcoded: [workspace-dashboard.js#L161](https://github.com/DataBiosphere/terra-ui/blob/adb4d9a1bdbc3a9d79033a2a63830b613665e302/integration-tests/tests/workspace-dashboard.js#L161) and [preview-drs-uri.js#L17](https://github.com/DataBiosphere/terra-ui/blob/adb4d9a1bdbc3a9d79033a2a63830b613665e302/integration-tests/tests/preview-drs-uri.js#L17) What do I do?
> > 
> > 
> > Those seem to mock API responses. @cahrens do they need to specify an environment?
> 
> @cahrens Hi, sorry for the re-asking because I don't find your answer to this question. Can you check it out?

I don't know anything about drs test, but as far as the `workspace-dashboard` one, I will follow up later on why the test passes against staging even though the `dev` billing project name is hardcoded. Don't hold up this PR as long as the tests pass.

**Comment by aweng98** [comment]
Two UI tests (`run-catalog-workflow.js` and `link-to-new-workspace.js`) failed due to `page.http GET 401 https://data.staging.envs-terra.bio/api/repository/v1/snapshots/62f9a3a4-0fb7-4694-b6b2-e1e65b022045/export?validatePrimaryKeyUniqueness=false`. Rerunning workflow to see if happens again.

**Comment by nawatts** [comment]
run-catalog and link-to-new-workspace are also currently failing on `dev`: https://app.circleci.com/pipelines/github/DataBiosphere/terra-ui?branch=dev

**Comment by nawatts** [comment]
Since the two tests that are failing on this PR are also failing on dev, assuming that they're unrelated.

**Review by nawatts — APPROVED** [review]

**Comment by aweng98** [comment]
@nawatts @breilly2 FYI, rebased to latest on Dev branch and all tests have passed this time.

**Review by nawatts — APPROVED** [review]

**Review by breilly2 — APPROVED** [review]

**Pull request merged**
